README.md
[![Travis](https://img.shields.io/travis/locize/locize-editor/master.svg?style=flat-square)](https://travis-ci.org/locize/locize-editor)
[![npm version](https://img.shields.io/npm/v/locize-editor.svg?style=flat-square)](https://www.npmjs.com/package/locize-editor)
[![Bower](https://img.shields.io/bower/v/locize-editor.svg)]()
[![David](https://img.shields.io/david/locize/locize-editor.svg?style=flat-square)](https://david-dm.org/locize/locize-editor)

# locize-editor

The locize-editor enables you to directly connect content from your website / application with your content on your localization project on locize.

Enabling the editor by querystring `?locize=true` you can click any text content on your website to open it on the right side editor window:

![](http://docs.locize.com/assets/editor.png 'Screen of InContext Editor')

The linking could be turned on/off using the button on the lower right or by pressing `ctrl-x`.

## Getting started

Source can be loaded via [npm](https://www.npmjs.com/package/locize-editor), bower, [downloaded](https://github.com/locize/locize/blob/master/locize.min.js) from this repo or loaded from the npm CDN [unpkg.com/locize-editor](https://unpkg.com/locize-editor/locize-editor.min.js).

If not using a bundler the script will be added to `window.locizeEditor`.

**Hint:** This module runs only in browser.

## Using

### standalone

Just init like:

```js
locizeEditor.init({
  lng: 'fr',
  defaultNS: 'namespaceToUse',
  projectId: '[yourProjectID]',
  referenceLng: 'en'
});
```

Open edit mode by appending `?locize=true` to the querystring.

For additional options see below **Initialize with optional options**

### with locizify

The editor is built into our [locizify script](https://github.com/locize/locizify). There is no additional step needed.

Open edit mode by appending `?locize=true` to the querystring.

For texts using plural or interpolation feature you might need to additionally add `&lng=cimode&useLng=[yourLocal]` to find a key.

### with i18next and i18next-locize-backend

```
import locizeEditor from 'locize-editor';
import i18next from 'locize';
import Backend from 'i18next-locize-backend';

i18next
  .use(Backend)
  .use(locizeEditor);
```

### Initialize with optional options

You can configure some aspects like layout by adding init options.

```js
// standalone
locizeEditor.init({
  // enable on init without the need of adding querystring locize=true
  enabled: false,
  autoOpen: true, // if set to false you will need to open it via API

  openDashboard: false, // open dashboard instead of the specified project

  // enable by adding querystring locize=true; can be set to another value or turned off by setting to false
  enableByQS: 'locize',

  // turn on/off by pressing
  toggleKeyModifier: 'ctrlKey', // metaKey | altKey | shiftKey
  toggleKeyCode: 24, // x when pressing ctrl (e.which: document.addEventListener('keypress', (e) => console.warn(e.which, e));

  // use lng in editor taken from query string, eg. if running with lng=cimode (i18next, locize)
  lngOverrideQS: 'useLng',

  // use lng in editor, eg. if running with lng=cimode (i18next, locize)
  lngOverride: null,

  // expect keys to have namespace prepend (CIMODE): text clicked is in form ns:key
  hasNamespacePrefixed: false,

  // default will open a iframe; setting to window will open a new window/tab instead
  mode: 'iframe', // 'window',

  // styles to adapt layout in iframe mode to your website layout
  iframeContainerStyle: 'z-index: 1000; position: fixed; top: 0; right: 0; bottom: 0; width: 600px; box-shadow: -3px 0 5px 0 rgba(0,0,0,0.5);',
  iframeStyle: 'height: 100%; width: 600px; border: none;',
  bodyStyle: 'margin-right: 605px;',

  // handle when locize saved the edited translations, eg. reload website
  // You may need to click first one of the text elements in your app first. (https://github.com/locize/locize-editor/issues/17)
  onEditorSaved: function(lng, ns) { location.reload(); },
  
  // custom appendTarget and container classes for the InContext Editor
  locizeEditorWindow: {
    appendTarget: document.querySelector('.example-div'), // Element into which the editor window should be appended. Element should be set position: relative as the editor window itself will be set position: absolute;. defaults to false, appending to the body
    containerClasses: 'custom-class second-custom-class' // defaults to false, resulting in no custom classes set
  },
  // custom appendTarget, container classes and title for the "locize editor" overlay with the button to toggle click selection of strings.
  locizeEditorToggle: {
    appendTarget: document.querySelector('.example-div'), // Element into which the toggle should be appended. Element should be set position: relative as the toggle itself will be set position: absolute;. defaults to false, appending to the body
    containerClasses: 'custom-class second-custom-class', // defaults to false, resulting in no custom classes set
    title: `Custom title` // defaults to false, resulting in the default title being displayed
  }
})

// i18next, ...
locizify|locize|i18next.init({
  editor: {
    // all options above
  }
});
```

### update your application when you saved changes in locize editor

use the onEditorSaved handler

```js
// reload the full page
locizeEditor.init({
  onEditorSaved: function(lng, ns) {
    location.reload();
  }
});

// reload translations in i18next and trigger a rerender
locizeEditor.init({
  onEditorSaved: function(lng, ns) {
    i18next.reloadResources(lng, ns, () => {
      // trigger rerender on your app if needed
      // note: reloadResources can take an optional callback in i18next@>=11.9.0
    });
  }
});
```

### update the locize editor if your site sent new save missings

the app will reload those namespaces

```js
locizeEditor.handleSavedMissing(lng, ns);

// eg. on your i18next-locize-backend bind the onSaved on init
i18next
  .use(locizeEditor)
  .use(locizeBackend)
  .init({
    //...
    backend: {
      //...
      onSaved: locizeEditor.handleSavedMissing
    }
  });
```

see the sample for react.js [i18next/react-i18next sample](https://github.com/i18next/react-i18next/blob/master/example/locize-example/src/i18n.js#L45)

### find the namespace

As content is structured into multiple translation files we somehow need to detect/find the namespace (file) used for translating the clicked content.

If locize is not able to detect the right namespace it will fallback for a fuzzy global search - which in most cases should also provide the correct result. But you might like to get more control.

##### by adding attribute containing namespace

The namespace will be detected from current clicked element or any of its parents.

Following attributes are valid to look it up:

```html
<!-- standalone -->
<div i18n-ns="myNamespace">content</div>
<div data-i18n-ns="myNamespace">content</div>

<!-- i18next: preferred -->
<div i18next-ns="myNamespace">content</div>
<div data-i18next-ns="myNamespace">content</div>

<!-- alternative: json stringified i18next options-->
<div i18next-options="{\"ns\":\"myNamespace\"}">content</div>
<div data-i18next-options="{\"ns\":\"myNamespace\"}">content</div>
```

##### by using cimode language

If there is no possibility to find the used namespace per attribute on parent element you will need to toggle the lng to cimode. Additionally you will need to configure locize to append namespace in cimode by:

```
i18next.init({
  appendNamespaceToCIMode: true
});
```

open your website with querystring `?locize=true&lng=cimode&useLng=[yourLocal]`.


CHANGELOG.md
### 3.1.1

- using explictly the locize domain for old client

### 3.1.0

- Customizable placement of editor / toggle window [22](https://github.com/locize/locize-editor/pull/22)

### 3.0.2

- fix when using with SSR

### 3.0.1

- auto-open editor on toggleUI

### 3.0.0

- using the new `locize.app` domain replacing the deprecated `locize.io` domain

### 2.2.2

- z-index correction on the toggle UI

### 2.2.1

- fix hasNamespacePrefixed

### 2.2.0

- only default namespace separator to ':' if undefined in i18next.options -> do not split : away on text if not set so
- add option hasNamespacePrefixed (default false) -> setting false will avoid splitting
- extend iframe z-index to 2147483647 based on [stackoverflow](https://stackoverflow.com/questions/491052/minimum-and-maximum-value-of-z-index/25461690) the max
- have a handler for handleSavedMissing(lng, ns)

### 2.1.3

- bind on open

### 2.1.2

- alternative search for projectId

### 2.1.0

- open project specified instead of dashboard

### 2.0.0

- removes deprecated jsnext:main from package.json
- Bundle all entry points with rollup [10](https://github.com/locize/locize-editor/pull/10)
- **note:** dist/es -> dist/esm, dist/commonjs -> dist/cjs (individual files -> one bundled file)
- removes bower finally

### v1.7.0

- allow setting lngOverride via init options as alternative to querystring [8](https://github.com/locize/locize-editor/pull/8)

### v1.6.3

- ignore ignorelocizeeditor: another instance [7](https://github.com/locize/locize-editor/pull/7)

### v1.6.2

- click handler: make ignorelocizeeditor elements clickable again [6](https://github.com/locize/locize-editor/pull/6)

### v1.6.1

- click event handler: prioritize over existing dom click events [5](https://github.com/locize/locize-editor/pull/5)

### v1.6.0

- option to trigger rerender on locize saved segments options.onEditorSaved: function(lng, ns) {}

### v1.5.1

- prevent link following [#4](https://github.com/locize/locize-editor/pull/4)

### v1.5.0

- support for input.placeholder

### v1.4.0

- standalone option to init without i18next usage

### v1.3.0

- read namespace from data-i18next-ns or i18next-ns too

### v1.2.0

- read namespace from data-i18next-options too

### v1.1.0

- enables listening to editor keypress on iframe - to toggle on/off while focus is on the iframe

### v1.0.0

- optimize z-index

### v0.12.0

- override lng with qs param useLng (useful for cimode usage)
- preventDefault on clicked element avoiding href navigation
- toggle on/off by keyboard ctrl+x (customizable)

### v0.11.0

- adds new default mode 'iframe', mode could be toggled by ?locizeMode=tab or setting mode on editor options

### v0.10.1

- replace cimode lng to referenceLng

### v0.10.0

- fixes autoOpen false
- removes html comments from detectable (enables usage eg. for react apps)
- remove namespace from submitted token

### v0.9.1

- adapt module.type change in i18next

### v0.9.0

- editor UI to toggle on/off
- enable by qs

### v0.0.1

- initial version


src/ui.js
const baseBtn = 'font-family: "Helvetica", "Arial", sans-serif; font-size: 14px; color: #fff; border: none; font-weight: 300; height: 30px; line-height: 30px; padding: 0; text-align: center; min-width: 90px; text-decoration: none; text-transform: uppercase; text-overflow: ellipsis; white-space: nowrap; outline: none; cursor: pointer;';

export function initUI(on, off, options) {
  const cont =  document.createElement("div");
  let style = 'font-family: "Helvetica", "Arial", sans-serif; bottom: 20px; right: 20px; padding: 10px; background-color: #fff; border: solid 1px #1976d2; box-shadow: 0px 1px 2px 0px rgba(0,0,0,0.5);';
  if(options.locizeEditorToggle.appendTarget) {
    style += 'position: absolute;';
  } else {
    style += ' z-index: 2147483647; position: fixed;';
  }
  cont.setAttribute('style', style);
  cont.setAttribute('ignorelocizeeditor', '');
  cont.setAttribute('translated', '');
  if(options.locizeEditorToggle.containerClasses) {
    const classes = options.locizeEditorToggle.containerClasses.length > 1 ? options.locizeEditorToggle.containerClasses.split(' ') : options.locizeEditorToggle.containerClasses;
    classes.forEach(function(cssClass) {
      cont.classList.add(cssClass);
    });
  }

  const title = document.createElement("h4");

  title.id = "locize-title";
  title.innerHTML = options.locizeEditorToggle.title || "locize editor";
  title.setAttribute('style', 'font-family: "Helvetica", "Arial", sans-serif; font-size: 14px; margin: 0 0 5px 0; color: #1976d2; font-weight: 300;')
  title.setAttribute('ignorelocizeeditor', '');
  cont.appendChild(title);

  const turnOff =  document.createElement("button");
  turnOff.innerHTML = "On";
  turnOff.setAttribute('style', `${baseBtn} display: none; background-color: #54A229;`)
  turnOff.onclick = off;
  turnOff.setAttribute('ignorelocizeeditor', '');
  cont.appendChild(turnOff);

  const turnOn =  document.createElement("button");
  turnOn.innerHTML = "Off";
  turnOn.setAttribute('style', `${baseBtn} display: none; background-color: #D50000;`)
  turnOn.onclick = on;
  turnOn.setAttribute('ignorelocizeeditor', '');
  cont.appendChild(turnOn);

  if(options.locizeEditorToggle.appendTarget) {
    options.locizeEditorToggle.appendTarget.appendChild(cont);
  } else {
    document.body.appendChild(cont);
  }

  const toggle = (on) => {
    turnOff.style.display = on ? 'block' : 'none';
    turnOn.style.display = !on ? 'block' : 'none';
  }

  return toggle;
}

export function appendIframe(url, options) {
  if(options.locizeEditorWindow.appendTarget) {
    options.iframeContainerStyle = 'position: absolute; top: 0; right: 0; bottom: 0; left: 0;';
    options.iframeStyle += ' width: 100%;';
  }
  const cont =  document.createElement("div");
  cont.setAttribute('style', options.iframeContainerStyle);
  cont.setAttribute('ignorelocizeeditor', '');
  cont.setAttribute('translated', '');
  if(options.locizeEditorWindow.containerClasses) {
    const classes = options.locizeEditorWindow.containerClasses.length > 1 ? options.locizeEditorWindow.containerClasses.split(' ') : options.locizeEditorWindow.containerClasses;
    classes.forEach(function(cssClass) {
      cont.classList.add(cssClass);
    });
  }

  const iframe =  document.createElement("iframe");
  iframe.setAttribute('style', options.iframeStyle);
  iframe.setAttribute('ignorelocizeeditor', '');
  iframe.setAttribute('translated', '');
  iframe.setAttribute('src', url);
  cont.appendChild(iframe);
  if(options.locizeEditorWindow.appendTarget) {
    options.locizeEditorWindow.appendTarget.appendChild(cont);
  } else {
    document.body.appendChild(cont);
    const bodyStyle = document.body.getAttribute('style');
    document.body.setAttribute('style', "".concat(bodyStyle, "; ").concat(options.bodyStyle));
  }
  return iframe.contentWindow;
}


src/utils.js
export function isWindow(obj) {
  return obj != null && obj === obj.window;
}

export function getWindow(elem) {
    return isWindow(elem) ? elem : elem.nodeType === 9 && elem.defaultView;
}

export function offset(elem) {
  var docElem, win,
      box = { top: 0, left: 0, right: 0, bottom: 0 },
      doc = elem && elem.ownerDocument;

  docElem = doc && doc.documentElement;
	if (!docElem) return box;

  if (typeof elem.getBoundingClientRect !== typeof undefined) {
    box = elem.getBoundingClientRect();
  }
  win = getWindow(doc);

  const top = box.top + win.pageYOffset - docElem.clientTop;
  const left = box.left + win.pageXOffset - docElem.clientLeft;
  return {
    top,
    left,
    right: left + (box.right - box.left),
    bottom: top + (box.bottom - box.top)
  };
};

export function getClickedElement(e) {
  // clicked input
  if (e.srcElement && e.srcElement.nodeType === 1) {
    if (e.srcElement.getAttribute && e.srcElement.getAttribute('ignorelocizeeditor') === '') return null;
    return e.srcElement;
  }

  let el, toHigh, toLeft, toLeftNextOffset;

  if (e.originalEvent && e.originalEvent.explicitOriginalTarget) {
    el = e.originalEvent.explicitOriginalTarget;
  } else {
    let parent = e.srcElement;
    if (parent.getAttribute && parent.getAttribute('ignorelocizeeditor') === '') return null;

    let left = e.pageX;
    let top = e.pageY;
    let pOffset = offset(parent);
    // console.warn('click', top, left);
    // console.warn('parent', parent, pOffset, parent.clientHeight, parent.offsetHeight);

    let topStartsAt = 0;
    let topBreaksAt;
    for (let i = 0; i < parent.childNodes.length; i++) {
      let n = parent.childNodes[i];
      let nOffset = offset(n);
      // console.warn('child', n, nOffset, n.clientHeight, n.offsetHeight)

      // if a node is with the bottom over the top click set the next child as start index
      if (n.nodeType === 1 && nOffset.bottom < top) topStartsAt = i + 1;

      // if node is below top click set end index to this node
      if (!topBreaksAt && nOffset.top + (n.clientHeight || 0) > top) topBreaksAt = i;
    }

    // check we are inside children lenght
    if (topStartsAt + 1 > parent.childNodes.length) topStartsAt = parent.childNodes.length - 1;
    if (!topBreaksAt) topBreaksAt = parent.childNodes.length;
    // console.warn('bound', topStartsAt, topBreaksAt)

    // inside our boundaries check when left is to big and out of clicks left
    for (let y = topStartsAt; y < topBreaksAt; y++) {
      let n = parent.childNodes[y];
      let nOffset = offset(n);

      if (!toLeft && nOffset.left > left) {
        break;
      }

      if (n && n.nodeType !== 8) el = n;
    }
  }
  return el;
}

export function removeNamespace(str, i18next) {
  let res = str;
  const nsSeparator = i18next.options.nsSeparator !== undefined ?  i18next.options.nsSeparator : ':';

  if (str.indexOf(nsSeparator) > -1) {
    const p = str.split(nsSeparator);
		p.shift();
		res = p.join(nsSeparator);
  }
	return res;
}

export function getElementNamespace(str, el, i18next) {
  let namespace = i18next.options.defaultNS;
  const nsSeparator = i18next.options.nsSeparator || ':';

  if (str.indexOf(nsSeparator) > -1) {
    namespace = str.split(nsSeparator)[0];
  } else {
    let found;

    const find = (el) => {
      let opts = el.getAttribute && el.getAttribute('i18next-options')
      if (!opts) opts = el.getAttribute && el.getAttribute('data-i18next-options');
      if (!opts) opts = el.getAttribute && el.getAttribute('i18n-options');
      if (!opts) opts = el.getAttribute && el.getAttribute('data-i18n-options');
      if (opts) {
        let jsonData = {};
        try {
          jsonData = JSON.parse(opts);
        } catch (e) {
          // not our problem here in editor
        }
        if (jsonData.ns) found = jsonData.ns
      }
      if (!found) found = el.getAttribute && el.getAttribute('i18next-ns');
      if (!found) found = el.getAttribute && el.getAttribute('data-i18next-ns');
      if (!found) found = el.getAttribute && el.getAttribute('i18n-ns');
      if (!found) found = el.getAttribute && el.getAttribute('data-i18n-ns');
      if (!found && el.parentElement) find(el.parentElement);
    }
    find(el);

    if (found) namespace = found;
  }

  return namespace;
}

export function getQueryVariable(variable) {
  // dont check query on SSR render
  if (typeof window === 'undefined') {
    return false;
  }
  
  const query = window.location.search.substring(1);
  const vars = query.split('&');
  for (var i = 0; i < vars.length; i++) {
    const pair = vars[i].split('=');
    if (pair[0] == variable) {
      return pair[1];
    }
  }
  return false;
}


locize-editor.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
  typeof define === 'function' && define.amd ? define(factory) :
  (global = global || self, global.locizeEditor = factory());
}(this, function () { 'use strict';

  function _defineProperty(obj, key, value) {
    if (key in obj) {
      Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
      });
    } else {
      obj[key] = value;
    }

    return obj;
  }

  function _objectSpread(target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i] != null ? Object(arguments[i]) : {};
      var ownKeys = Object.keys(source);

      if (typeof Object.getOwnPropertySymbols === 'function') {
        ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) {
          return Object.getOwnPropertyDescriptor(source, sym).enumerable;
        }));
      }

      ownKeys.forEach(function (key) {
        _defineProperty(target, key, source[key]);
      });
    }

    return target;
  }

  function _typeof(obj) {
    "@babel/helpers - typeof";

    if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
      _typeof = function _typeof(obj) {
        return typeof obj;
      };
    } else {
      _typeof = function _typeof(obj) {
        return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
      };
    }

    return _typeof(obj);
  }

  function isWindow(obj) {
    return obj != null && obj === obj.window;
  }
  function getWindow(elem) {
    return isWindow(elem) ? elem : elem.nodeType === 9 && elem.defaultView;
  }
  function offset(elem) {
    var docElem,
        win,
        box = {
      top: 0,
      left: 0,
      right: 0,
      bottom: 0
    },
        doc = elem && elem.ownerDocument;
    docElem = doc && doc.documentElement;
    if (!docElem) return box;

    if (_typeof(elem.getBoundingClientRect) !== ("undefined")) {
      box = elem.getBoundingClientRect();
    }

    win = getWindow(doc);
    var top = box.top + win.pageYOffset - docElem.clientTop;
    var left = box.left + win.pageXOffset - docElem.clientLeft;
    return {
      top: top,
      left: left,
      right: left + (box.right - box.left),
      bottom: top + (box.bottom - box.top)
    };
  }
  function getClickedElement(e) {
    // clicked input
    if (e.srcElement && e.srcElement.nodeType === 1) {
      if (e.srcElement.getAttribute && e.srcElement.getAttribute('ignorelocizeeditor') === '') return null;
      return e.srcElement;
    }

    var el;

    if (e.originalEvent && e.originalEvent.explicitOriginalTarget) {
      el = e.originalEvent.explicitOriginalTarget;
    } else {
      var parent = e.srcElement;
      if (parent.getAttribute && parent.getAttribute('ignorelocizeeditor') === '') return null;
      var left = e.pageX;
      var top = e.pageY;
      var pOffset = offset(parent); // console.warn('click', top, left);
      // console.warn('parent', parent, pOffset, parent.clientHeight, parent.offsetHeight);

      var topStartsAt = 0;
      var topBreaksAt;

      for (var i = 0; i < parent.childNodes.length; i++) {
        var n = parent.childNodes[i];
        var nOffset = offset(n); // console.warn('child', n, nOffset, n.clientHeight, n.offsetHeight)
        // if a node is with the bottom over the top click set the next child as start index

        if (n.nodeType === 1 && nOffset.bottom < top) topStartsAt = i + 1; // if node is below top click set end index to this node

        if (!topBreaksAt && nOffset.top + (n.clientHeight || 0) > top) topBreaksAt = i;
      } // check we are inside children lenght


      if (topStartsAt + 1 > parent.childNodes.length) topStartsAt = parent.childNodes.length - 1;
      if (!topBreaksAt) topBreaksAt = parent.childNodes.length; // console.warn('bound', topStartsAt, topBreaksAt)
      // inside our boundaries check when left is to big and out of clicks left

      for (var y = topStartsAt; y < topBreaksAt; y++) {
        var _n = parent.childNodes[y];

        var _nOffset = offset(_n);

        if (_nOffset.left > left) {
          break;
        }

        if (_n && _n.nodeType !== 8) el = _n;
      }
    }

    return el;
  }
  function removeNamespace(str, i18next) {
    var res = str;
    var nsSeparator = i18next.options.nsSeparator !== undefined ? i18next.options.nsSeparator : ':';

    if (str.indexOf(nsSeparator) > -1) {
      var p = str.split(nsSeparator);
      p.shift();
      res = p.join(nsSeparator);
    }

    return res;
  }
  function getElementNamespace(str, el, i18next) {
    var namespace = i18next.options.defaultNS;
    var nsSeparator = i18next.options.nsSeparator || ':';

    if (str.indexOf(nsSeparator) > -1) {
      namespace = str.split(nsSeparator)[0];
    } else {
      var found;

      var find = function find(el) {
        var opts = el.getAttribute && el.getAttribute('i18next-options');
        if (!opts) opts = el.getAttribute && el.getAttribute('data-i18next-options');
        if (!opts) opts = el.getAttribute && el.getAttribute('i18n-options');
        if (!opts) opts = el.getAttribute && el.getAttribute('data-i18n-options');

        if (opts) {
          var jsonData = {};

          try {
            jsonData = JSON.parse(opts);
          } catch (e) {// not our problem here in editor
          }

          if (jsonData.ns) found = jsonData.ns;
        }

        if (!found) found = el.getAttribute && el.getAttribute('i18next-ns');
        if (!found) found = el.getAttribute && el.getAttribute('data-i18next-ns');
        if (!found) found = el.getAttribute && el.getAttribute('i18n-ns');
        if (!found) found = el.getAttribute && el.getAttribute('data-i18n-ns');
        if (!found && el.parentElement) find(el.parentElement);
      };

      find(el);
      if (found) namespace = found;
    }

    return namespace;
  }
  function getQueryVariable(variable) {
    // dont check query on SSR render
    if (typeof window === 'undefined') {
      return false;
    }

    var query = window.location.search.substring(1);
    var vars = query.split('&');

    for (var i = 0; i < vars.length; i++) {
      var pair = vars[i].split('=');

      if (pair[0] == variable) {
        return pair[1];
      }
    }

    return false;
  }

  var baseBtn = 'font-family: "Helvetica", "Arial", sans-serif; font-size: 14px; color: #fff; border: none; font-weight: 300; height: 30px; line-height: 30px; padding: 0; text-align: center; min-width: 90px; text-decoration: none; text-transform: uppercase; text-overflow: ellipsis; white-space: nowrap; outline: none; cursor: pointer;';
  function initUI(on, off, options) {
    var cont = document.createElement("div");
    var style = 'font-family: "Helvetica", "Arial", sans-serif; bottom: 20px; right: 20px; padding: 10px; background-color: #fff; border: solid 1px #1976d2; box-shadow: 0px 1px 2px 0px rgba(0,0,0,0.5);';

    if (options.locizeEditorToggle.appendTarget) {
      style += 'position: absolute;';
    } else {
      style += ' z-index: 2147483647; position: fixed;';
    }

    cont.setAttribute('style', style);
    cont.setAttribute('ignorelocizeeditor', '');
    cont.setAttribute('translated', '');

    if (options.locizeEditorToggle.containerClasses) {
      var classes = options.locizeEditorToggle.containerClasses.length > 1 ? options.locizeEditorToggle.containerClasses.split(' ') : options.locizeEditorToggle.containerClasses;
      classes.forEach(function (cssClass) {
        cont.classList.add(cssClass);
      });
    }

    var title = document.createElement("h4");
    title.id = "locize-title";
    title.innerHTML = options.locizeEditorToggle.title || "locize editor";
    title.setAttribute('style', 'font-family: "Helvetica", "Arial", sans-serif; font-size: 14px; margin: 0 0 5px 0; color: #1976d2; font-weight: 300;');
    title.setAttribute('ignorelocizeeditor', '');
    cont.appendChild(title);
    var turnOff = document.createElement("button");
    turnOff.innerHTML = "On";
    turnOff.setAttribute('style', "".concat(baseBtn, " display: none; background-color: #54A229;"));
    turnOff.onclick = off;
    turnOff.setAttribute('ignorelocizeeditor', '');
    cont.appendChild(turnOff);
    var turnOn = document.createElement("button");
    turnOn.innerHTML = "Off";
    turnOn.setAttribute('style', "".concat(baseBtn, " display: none; background-color: #D50000;"));
    turnOn.onclick = on;
    turnOn.setAttribute('ignorelocizeeditor', '');
    cont.appendChild(turnOn);

    if (options.locizeEditorToggle.appendTarget) {
      options.locizeEditorToggle.appendTarget.appendChild(cont);
    } else {
      document.body.appendChild(cont);
    }

    var toggle = function toggle(on) {
      turnOff.style.display = on ? 'block' : 'none';
      turnOn.style.display = !on ? 'block' : 'none';
    };

    return toggle;
  }
  function appendIframe(url, options) {
    if (options.locizeEditorWindow.appendTarget) {
      options.iframeContainerStyle = 'position: absolute; top: 0; right: 0; bottom: 0; left: 0;';
      options.iframeStyle += ' width: 100%;';
    }

    var cont = document.createElement("div");
    cont.setAttribute('style', options.iframeContainerStyle);
    cont.setAttribute('ignorelocizeeditor', '');
    cont.setAttribute('translated', '');

    if (options.locizeEditorWindow.containerClasses) {
      var classes = options.locizeEditorWindow.containerClasses.length > 1 ? options.locizeEditorWindow.containerClasses.split(' ') : options.locizeEditorWindow.containerClasses;
      classes.forEach(function (cssClass) {
        cont.classList.add(cssClass);
      });
    }

    var iframe = document.createElement("iframe");
    iframe.setAttribute('style', options.iframeStyle);
    iframe.setAttribute('ignorelocizeeditor', '');
    iframe.setAttribute('translated', '');
    iframe.setAttribute('src', url);
    cont.appendChild(iframe);

    if (options.locizeEditorWindow.appendTarget) {
      options.locizeEditorWindow.appendTarget.appendChild(cont);
    } else {
      document.body.appendChild(cont);
      var bodyStyle = document.body.getAttribute('style');
      document.body.setAttribute('style', "".concat(bodyStyle, "; ").concat(options.bodyStyle));
    }

    return iframe.contentWindow;
  }

  var defaultOptions = {
    url: 'https://old.locize.app',
    openDashboard: false,
    enabled: false,
    enableByQS: 'locize',
    toggleKeyCode: 24,
    toggleKeyModifier: 'ctrlKey',
    lngOverrideQS: 'useLng',
    lngOverride: null,
    hasNamespacePrefixed: false,
    autoOpen: true,
    onEditorSaved: function onEditorSaved(lng, ns) {},
    mode: getQueryVariable('locizeMode') || 'iframe',
    iframeContainerStyle: 'z-index: 2147480000; position: fixed; top: 0; right: 0; bottom: 0; width: 600px; box-shadow: -3px 0 5px 0 rgba(0,0,0,0.5);',
    iframeStyle: 'height: 100%; width: 600px; border: none;',
    bodyStyle: 'margin-right: 605px;',
    locizeEditorWindow: {
      appendTarget: false,
      containerClasses: false
    },
    locizeEditorToggle: {
      appendTarget: false,
      containerClasses: false,
      title: false
    }
  };

  function convertOptionsToI18next(opts) {
    return {
      languages: [opts.lng],
      nsSeparator: opts.nsSeparator || ':',
      options: {
        editor: opts,
        backend: opts,
        defaultNS: opts.defaultNS
      }
    };
  }

  var editor = {
    type: '3rdParty',
    init: function init(i18next) {
      var _this = this;

      // convert standalone options
      if (i18next && !i18next.init) i18next = convertOptionsToI18next(i18next);
      this.enabled = false;
      this.i18next = i18next;
      this.options = _objectSpread({}, defaultOptions, i18next.options.editor); //this.locizeUrl = (i18next.options.editor && i18next.options.editor.url) || 'https://old.locize.app';

      this.handler = this.handler.bind(this);
      this.handleSavedMissing = this.handleSavedMissing.bind(this);

      if (this.options.enabled || this.options.enableByQS && getQueryVariable(this.options.enableByQS) === 'true') {
        setTimeout(function () {
          if (_this.options.autoOpen) _this.open();
        }, 500);
      }

      document.addEventListener('keypress', function (e) {
        if (e[_this.options.toggleKeyModifier] && e.which === _this.options.toggleKeyCode) _this.enabled ? _this.off() : _this.on();
      }); // listen to key press on locize service to disable

      window.addEventListener('message', function (e) {
        if (e.data[_this.options.toggleKeyModifier] && e.data.which === _this.options.toggleKeyCode) _this.enabled ? _this.off() : _this.on();

        if (e.data.type === 'savedSegments') {
          _this.options.onEditorSaved(e.data.lng, e.data.ns);
        }
      });
    },
    handler: function handler(e) {
      var _this2 = this;

      var el = getClickedElement(e);
      if (!el) return;
      e.preventDefault();
      e.stopPropagation();
      var str = el.textContent || el.text && el.text.innerText || el.placeholder;
      if (typeof str !== 'string') return;
      var res = str.replace(/\n +/g, '').trim();

      var send = function send() {
        // consume
        // window.addEventListener('message', function(ev) {
        //   if (ev.data.message === 'searchForKey') {
        //     console.warn(ev.data);
        //   }
        // });
        var payload = {
          message: 'searchForKey',
          projectId: _this2.i18next.options.backend.projectId,
          version: _this2.i18next.options.backend.version || 'latest',
          lng: getQueryVariable(_this2.options.lngOverrideQS) || _this2.options.lngOverride || _this2.i18next.languages[0],
          ns: getElementNamespace(res, el, _this2.i18next),
          token: _this2.options.hasNamespacePrefixed ? removeNamespace(res, _this2.i18next) : res
        };
        if (!payload.lng || payload.lng.toLowerCase() === 'cimode') payload.lng = _this2.i18next.options.backend.referenceLng;
        if (_this2.options.handler) return _this2.options.handler(payload);

        _this2.locizeInstance.postMessage(payload, _this2.options.url);

        _this2.locizeInstance.focus();
      }; // assert the locizeInstance is still open


      if (this.options.autoOpen && (this.options.mode !== 'iframe' && !this.locizeInstance || this.locizeInstance.closed)) {
        this.open();
        setTimeout(function () {
          send();
        }, 3000);
      } else {
        send();
      }
    },
    handleSavedMissing: function handleSavedMissing(lng, ns) {
      if (!this.locizeInstance || this.locizeInstance.closed) return;
      var payload = {
        message: 'savedMissings',
        projectId: this.i18next.options.backend.projectId,
        version: this.i18next.options.backend.version || 'latest',
        lng: lng,
        ns: ns
      };
      this.locizeInstance.postMessage(payload, this.options.url);
    },
    open: function open() {
      var url = this.options.url;
      if (!this.options.openDashboard) url = "".concat(url, "/pid/").concat(this.options.projectId || this.i18next.options.backend.projectId, "/v/").concat(this.i18next.options.backend.version || 'latest');

      if (this.options.mode === 'iframe') {
        this.locizeInstance = appendIframe(url, this.options);
      } else {
        this.locizeInstance = window.open(url);
      } // bind toggle UI


      this.toggleUI = initUI(this.on.bind(this), this.off.bind(this), this.options); // start listening

      this.on();
    },
    on: function on() {
      document.body.addEventListener('click', this.handler, true);
      if (!this.toggleUI) this.open();
      this.toggleUI(true);
      this.enabled = true;
    },
    off: function off() {
      document.body.removeEventListener('click', this.handler, true);
      if (!this.toggleUI) this.open();
      this.toggleUI(false);
      this.enabled = false;
    }
  };

  return editor;

}));
